### ISSUE TITLE
enqueued mutation related to apollo-upload-client

### ISSUE THREAD

**Issue by stydav** [aerogear/offix#458]
* **Affected Package(s)**: offix-client
* **Package version(s)**: 0.15.1
* **Platform (e.g. Android/iOS)**: Android
* **Platform Version**: 8.0.0
* **Node.js / npm versions**: 12.14.1

when an image upload is triggered while offline, the operation gets enqueued. The mutation is like this:

`mutation UploadMutation($file: Upload!){
    uploadMutation(file: $file){
        url
    }
}`

apollo-upload-client version is 13.0.0. Uploading while online works fine so no compatibility issues there.

the offline queue operation fails because offix sends the upload mutation in exactly the same format as apollo would send it without having the upload link from apollo-upload-client set up (results in a non-uploadable file).

So this file transformation piece that apollo upload client does gets lost somehow. 

Is there a forseen way to deal with this kind of thing?
Otherwise grateful for any recommandations or workarounds on how to use the current functionality to deal with this kind of issue.

**Comment by wtrocki** [comment]
Yes. We initially had that upload client configured in our sample application and even offix had that as option embedded. But since this brought the package size we decided to remove it from the core package and we are offering offix-boost that have this capability:


What you might be having issue with is the HTTP link were not swapped with the FileUpload HttpLink.
See:

https://github.com/aerogear/offix/blob/master/packages/offix-client-boost/src/links/index.ts#L41

I will try to replicate it in the sample app in the coming days and come back to yourself with why this happens. The protocol is interpreted by HttpLink and offix acts before it so technically all should work.

**Comment by wtrocki** [comment]
CC @darahayes

**Comment by stydav** [comment]
that would be great thanks

**Comment by wtrocki** [comment]
@stydav would using offix boost resolve this issue for you. 
I have used offix boost as faster way to try that out and were able to send file (I did not had handler on backend though.

**Comment by stydav** [comment]
file is not being transformed with offix boost as offix-client with manually initialized apollo-upload-link does, this can be seen when logging the file on the backend. Ill provide an example on the bottom.

tried the following with the same codebase:
  1.) offix-client apollo-upload-link (ApolloLink)
  2.) offix-client-boost with `fileUpload: true`

the first transforms the file, upload works, the second does not, see the example
for offix-boost it is the same result as if no upload link had been initialized, i.e. the file transformation does not occur

but even if this worked, the issue also related to the queue: the queued operation, when going back online, should go through this apollo link chain in order for the file transformation to occur

is this currently the case for the queued operations?

---------------------------------------------------------------------------------
example for a) "raw" and b) transformed file

a) offix boost (which to my knowledge also uses apollo-upload-client as set in 'packages\offix-client-boost\src\links\index.ts' line 43): `<ActionController::Parameters {"uri"=>"file:///data/user/0/host.exp.exponent/cache/ExperienceData/%2540stydav%252Fteam_a/Camera/add77367-1311-46ea-90d1-b8070971114f.jpg", "name"=>"photo_1590415347823.jpg", "type"=>"image/jpg"} permitted: true>`

the file is initialized in the react native app as follows: `new ReactNativeFile({
    uri: imageUri,
    name: `photo_${+new Date()}.${fileType}`,
    type: `image/${fileType}`,
  })`

as it can bee seen that the file is not being transformed

b) offix-client with manually initialized apollo-upload-link: `<ActionDispatch::Http::UploadedFile:0x000000000adbb758 @tempfile=#<Tempfile:C:/Users/---/AppData/Local/Temp/RackMultipart20200525-16892-7jv8fo.jpg>, @original_filename="photo_1590415712125.jpg", @content_type="image/jpg", @headers="content-disposition: form-data; name=\"1\"; filename=\"photo_1590415712125.jpg\"\r\nContent-Type: image/jpg\r\nContent-Length: 716017\r\n">`

as it can be seen the file is being transformed

**Comment by wtrocki** [comment]
@stydav Yes. I will update the sample app to use boost to properly test it.

@kingsleyzissou WDYT?

**Comment by kingsleyzissou** [comment]
@wtrocki datasync starter would probably be a better shout.

**Comment by gaurav-tangelo** [comment]
any update here ? We are seeing the same issue. Thanks so much @wtrocki and @kingsleyzissou

**Comment by wtrocki** [comment]
Hi guys. Sorry for not comming back to you. 

TL;DR let's use offix-client for the moment as workaround and we going to fix offix boost in the next release which will should happen soon anyway

**Comment by gaurav-tangelo** [comment]
@wtrocki do you know approximately when will the next release with the fix be available ? Thanks in advance

**Comment by wtrocki** [comment]
while you can move to self configured client without boost now. We will release new version as soon this is fixed. Sometime next week

**Comment by wtrocki** [comment]
Working on this now.

**Comment by gaurav-tangelo** [comment]
thank you :)

**Comment by wtrocki** [comment]
Hi guys.. I have tried this in boost and this is simply due to wrong version. I think we just need to release new version - Using boost from source worked for me, but release did not. 
We will do more package updates and release offix next week

**Comment by stydav** [comment]
Awesome, thanks

**Comment by stydav** [comment]
Hi, just checking in, is this released?

**Comment by wtrocki** [comment]
Release happening today evening. small fixes required. Sorry

**Comment by stydav** [comment]
alright thanks, no problem

**Comment by stydav** [comment]
hi, ive looked at the npm versions and could didnt find a new release yet. When do you think this will (approximately) be ready, since we would really need it for our application

**Comment by wtrocki** [comment]
Hi
We releasing it now. Expect release in 10 min

**Comment by stydav** [comment]
hello, the latest releases unfortunately did not fix the error.

If an image is being uploaded while offline, a queue entry is created. When coming back online the image is not being sent throught the usual link chain as it would be if online. 

Steps to reproduce:

1. simulate being offline
2. trigger an image upload
3. go back online
4. see file on serverside -> not transformed

please give me feedback if you can reproduce and fix it

**Comment by wtrocki** [comment]
@stydav I have tried this. You have mentioned that you are using boost. 
If you use offix-client directly with the file-upload HTTP link is that working. 

I will reopen this and setup reproduction repo.

**Comment by stydav** [comment]
oh sorry must not have clarified that: as of right now were not using boost, it was tried to test out the case.

In any case, I think for the package to work stable the configured link chain (i.e. fileupload in this case) with and without using offix-boost and for both online and queued operations will be of use to the users

**Comment by stydav** [comment]
could you reproduce the queue usecase?

**Comment by wtrocki** [comment]
I finally managed to reproduce it now. I think we need to detect file and transform it properly as operation that we are building seems to be missing something.

**Comment by stydav** [comment]
great, yeah exactly
basically to send it throught the configured ApolloLink again
so in this case the transformation with the upload link

maybe there is a generic solution to it and it is similar as the core apollo client library does it

**Comment by wtrocki** [comment]
Generally, I found it tricky to restore as we need to bring the original operation that comes from scalar. 

> b) offix-client with manually initialized apollo-upload-link: <ActionDispatch::Http::UploadedFile:0x000000000adbb758 @tempfile=#<Tempfile:C:/Users

I see that you had that working with the classical offix and got confused. 
I will need to sit down for a day and get this fixed but currently busy with other issues.

Problem is happening here:
https://github.com/aerogear/offix/blob/11f1c107606733b17cc897b4e10ed38291d7f2d9/packages/offix-scheduler/src/queue/OfflineQueue.ts#L139

and it is related to operation not containing all required fields that file upload needs.

**Comment by stydav** [comment]
queued operations where unfortunately not working with any configuration so far

ok yeah no problem, we are looking for a workaround in the meantime, please keep me updated

thanks

**Comment by wtrocki** [comment]
OK I have solution. Generally when persisting file to offline queue it lands as object.

<img width="315" alt="Screenshot 2020-06-25 at 12 53 21" src="https://user-images.githubusercontent.com/981838/85715088-fa836280-b6e2-11ea-81d1-522e2e68b8f9.png">


We can reproduce that in #537

**Comment by wtrocki** [comment]
I could apply some ugly hack on the scheduler to get this done - to not use persisted version from the queue but actual file object. But this will still fail short. 

Reason why this worked for us that previous releases actually always used in memory FileList. 
I think that issue here is that fileUpload link issuing one of the fileList methods so if we wrap our file to have this method that will work for your specific use case in separate apollo link.

**Comment by wtrocki** [comment]
Further investigation I found that object we have in cache  needs to be just instance of File. 
After landing in queue it becomes just simple object.
This is the code that is tricky:
https://github.com/jaydenseric/apollo-upload-client/blob/26e4b37db2ad1df12e499a42dab04dfcabe5669c/src/helpers.js#L21-L23

For us to resolve this problem we will need to get some directive to mark the certain fields as File or detect file scalar (but that might not work when passing input types etc.

To enable file support we will need to build something like:

```
const middlewareLink = new ApolloLink((operation, forward) => {
  operation.variables.file = new FileWrapper(operation.variables.file)
  return forward(operation);
});
```

If your platform is Web file wrapper is not possible, however any other platform can work with this pretty nice.

**Comment by stydav** [comment]
ok mostly got whats the issue
as of right now we implemented a custom solution
but happy to use the package for it when its ready

thanks for your time!

**Comment by gaurav-tangelo** [comment]
@wtrocki do you think this will issue will be fixed in the near future. We are currently using the library and this is one of the biggest feature (offline picture upload) that is not working. Would be good to know the status on this one. Thanks in advance and appreciate your work.

**Comment by wtrocki** [comment]
@gaurav-tangelo I figured fix! All you need to do is:

```ts
operation.variables.yourFileProperty.__proto__ = File.prototype
```
This will trick upload link here:
https://github.com/jaydenseric/apollo-upload-client/blob/26e4b37db2ad1df12e499a42dab04dfcabe5669c/src/helpers.js#L21-L23

Persistence strip out prototype but actual properties of the file object is preserved
Entire link that should be added before upload client. 

```ts
const middlewareLink = new ApolloLink((operation, forward) => {
  operation.variables.yourFileProperty.__proto__ = File.prototype
  return forward(operation);
});
```

Sorry that this took so long.

**Comment by wtrocki** [comment]
I will see if we can add native support for this scalar.

**Comment by kingsleyzissou** [comment]
Hi @stydav, we have made the decision to move away from the Apollo client and we have decided to deprecate our Offix packages and release our Datastore. I will be closing this issue, but if you have any questions, please feel free to let us know.

You can see the docs for the updated datastore here:
https://offix.dev/docs/getting-started
